test.cpp

```diff
@@ -0,0 +1,180 @@
+	if(l.size() != 5){
+		return "Failed. Size test.";
+	}
+	l.delete_first();
+	l.delete_first();
+	if(l.size() != 3 || l[0] != value_2){
+		return "Failed. Delete_first test.";
+	}
+	l.delete_last();
+	l.delete_last();
+	if(l.size() != 1 || l[0] != value_2){
+		return "Failed. Delete_last test.";	
+	}
+	return "Passed";
+
+}
+
+template<typename T>
+std::string test_2(T value_1, T value_2){
+	int rval = 0;
+	data_structures::list<T> l;
+	if(l.is_empty()){
+		rval++;
+	}
+	l.push_back(value_1);
+	l.push_back(value_1);
+	l.push_back(value_2);
+	l.push_back(value_1);
+	l.push_back(value_1);
+	if(l.is_empty()){
+		rval--;
+	}
+	if(!l.is_empty()){
+		rval++;
+	}
+	if(rval != 2){
+		return "Failed. Empty test.";
+	}
+	l.delete_concrete(value_2);
+	if(l.size() != 4 || l[2] == value_2){
+		if(l.size() != 4){
+			return "Failed. Size.";
+		}
+		return "Failed. Delete concrete test.";
+	}
+	l.insert(1, value_2);
+	l.insert(3, value_2);
+	if(l.size() != 6 || l[1] != value_2 || l[3] != value_2){
+		if(l.size() != 6){
+			return "Failed. Size.";
+		}
+		return "Failed. Insert test.";	
+	}
+	T val_t;
+	l.insert(4, val_t);
+	if(l.size() != 7 || l[4] != val_t){
+		return "Failed. Find index test.";	
+	}
+	return "Passed";
+}
+
+/*template<typename T>
+std::string test_3(T value_1, T value_2, T value_3, T value_4, T value_5){
+	data_structures::list<T> l;
+	l.push_back(value_1);
+	l.push_back(value_2);
+	l.push_back(value_3);
+	l.push_back(value_4);
+	l.push_back(value_5);
+	data_structures::list<T> l2;
+	l2.push_back(value_5);
+	l2.push_back(value_4);
+	l2.push_back(value_3);
+	l2.push_back(value_2);
+	l2.push_back(value_1);
+	l = l2;
+	l.print();
+	if(l[3] != value_2){
+		return "Failed";
+	}
+	return "Passed";
+}*/
+
+template<typename T>
+std::string test_4(T value_1, T value_2, T value_3){
+	data_structures::list<T> l;
+	l.push_back(value_1);
+	l.push_back(value_2);
+	l.push_back(value_3);
+	l.push_back(value_2);
+	l.push_back(value_3);
+	std::cout << std::endl;
+	data_structures::Node<T>* rval = l.find(value_3);
+	std::cout << rval -> value << std::endl;
+	std::cout << l[2] << std::endl;
+	std::cout << l.find(value_3) << std::endl;
+	std::cout << &l[2] << std::endl;
+	return "Passed";
+}
+
+template<typename T>
+std::string test_5(T value_1, T value_2, T value_3, T value_4){
+	data_structures::list<T> l;
+	data_structures::list<T> list_2;
+	l.push_back(value_1);
+	l.push_back(value_2);
+	l.push_back(value_3);
+	list_2.push_back(value_4);
+	list_2.push_back(value_4);
+	list_2.push_back(value_4);
+	list_2.push_back(value_4);
+	list_2 = l;
+	if(list_2.size() == 3 && list_2[0] == value_1 && list_2[1] == value_2 &&list_2[2] == value_3){
+		return "Passed";
+	}
+	return "Failed";
+}
+
+int main(){
+
+	std::cout << std::endl;
+	std::cout << "Test 1." << std::endl;
+	std::cout << "Includes: size(), push_back(), delete_first(), delete_last()." << std::endl;
+	std::cout << std::endl;
+	std::cout << "test 1.1: "<< test_1(113, 34, 55) << std::endl;
```

Comment:
I'd suggest to mention exact `T` type for clearer output (may be helpful in case of failures).